### DISCUSSION TITLE
Why not to use backgroundworkers for heavy tasks?

### DISCUSSION THREAD

**Discussion by rohansaw** [fastapi/fastapi#6866] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description
I read that it is not advised to use fastapi's backgroundworker for heavy background tasks such as machine learning [Docs](https://fastapi.tiangolo.com/tutorial/background-tasks/#caveat). Could someone explain why exactly this is not advised?

Also: Would it be fine to start a new Process from a backgroundworker and do the heavy computation over there and just use the backgroundworker to wait for this process to terminate or would this to not be recommended?

**Comment by aalifadv** [comment]
Hi @rohansaw , 

> I read that it is not advised to use fastapi's backgroundworker for heavy background tasks such as machine learning Docs. Could someone explain why exactly this is not advised?

[Seperation of concern](https://en.wikipedia.org/wiki/Separation_of_concerns) is what came to my mind first.  Why would you have the application serving HTTP concerned with machine learning task? Only if you need give your machine learning task access to the same variable  & object that FastAPI has access to. Let alone the efficiency, realiability concerns.

> Also: Would it be fine to start a new Process from a backgroundworker and do the heavy computation over there and just use the backgroundworker to wait for this process to terminate or would this to not be recommended?

I would not recommend that. I am not sure if that would even work. You will be better off using [Celery](https://docs.celeryproject.org/en/stable/getting-started/introduction.html) which was designed to solve cases like yours.

**Comment by ycd** [comment]
Addition to @aalifadv comment, there is also https://github.com/encode/starlette/issues/919.

**Comment by Kludex** [comment, marked as answer]
> > Also: Would it be fine to start a new Process from a backgroundworker and do the heavy computation over there and just use the backgroundworker to wait for this process to terminate or would this to not be recommended?

> I would not recommend that. I am not sure if that would even work. You will be better off using Celery which was designed to solve cases like yours.

```python
from concurrent.futures import ProcessPoolExecutor

from fastapi import FastAPI

app = FastAPI()

executor = None

@app.on_event("startup")
async def startup():
    global executor
    executor = ProcessPoolExecutor()
    
@app.on_event("shutdown")
async def shutdown():
    global executor
    executor.shutdown()
```
It works. In some specific cases can be useful.

> I read that it is not advised to use fastapi's backgroundworker for heavy background tasks such as machine learning Docs. Could someone explain why exactly this is not advised?

It's about reliability and the retry mechanism. Background tasks are unreliable, if they fail, your system will not act accordingly. Besides, your API server is suppose to be lightweight and focus on receiving requests. Besides Celery, there's also [ARQ](https://github.com/samuelcolvin/arq) that can be used as an alternative.

**Comment by rohansaw** [comment]
> [Seperation of concern](https://en.wikipedia.org/wiki/Separation_of_concerns) is what came to my mind first. Why would you have the application serving HTTP concerned with machine learning task? Only if you need give your machine learning task access to the same variable & object that FastAPI has access to. Let alone the efficiency, realiability concerns.

Thanks @aalifadv for the answer! I have the case that I am building a webapp, where the user starts ML processes from the frontend and will then get live output and results from the ML-task that he started and the database gets updatet with the output of the ML processes. I didn't have experience yet with celery so I thaught I maybe could just implemented this like this but it seems like this is not the way to go

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @rohansaw  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
